### PR TITLE
[PATCH v7] linux-gen: sched: dynamic load balance

### PR DESCRIPTION
**Pull request by psavol**
Adds dynamic load balance of spread queues. It's enabled on default and can be disabled from config file.

### CODE REVIEW

File: platform/linux-generic/odp_schedule_basic.c [hunk 1]
```diff
@@ -881,6 +947,46 @@ static int schedule_config(const odp_schedule_config_t *config)
...
+	cur_load = spread_load(grp, prio, cur_spr, 0);
+	min_load = cur_load;
+
+	for (spr = 0; spr < num_spread; spr++) {
+		if (spr == cur_spr)
+			continue;
+
+		load = spread_load(grp, prio, spr, 1);
+
+		/* Move queue if improvement is larger than marginal */
```
Comment (MatiasElo) [inline platform/linux-generic/odp_schedule_basic.c:980]: Typo: margin

**Review by MatiasElo — APPROVED** [review]

File: platform/linux-generic/odp_schedule_basic.c [hunk 2]
```diff
@@ -357,6 +388,8 @@ static int read_config_file(sched_global_t *sched)
 	sched->config_if.group_enable.control = val;
 	ODP_PRINT("  %s: %i\n", str, val);
 
+	ODP_PRINT("  dynamic load balance: %s\n", sched->load_balance ? "ON" : "OFF");
```
[the remark this answers is not in the document]
Comment (psavol) [reply platform/linux-generic/odp_schedule_basic.c:391]: balance is OFF, when prio_spread is 1 or load_balance is 0.

[the remark this answers is not in the document]
Comment (MatiasElo) [reply platform/linux-generic/odp_schedule_basic.c:391]: ok

File: platform/linux-generic/odp_schedule_basic.c (hunk 1 above)
Comment (psavol) [reply platform/linux-generic/odp_schedule_basic.c:980]: Forgot to fix this, but it's actually not a typo. It means to say that: when improvement is less than "margin" it is marginal (not significant).

**Comment by psavol** [comment]
v3: rebased

File: platform/linux-generic/odp_schedule_basic.c [hunk 3]
```diff
@@ -205,9 +205,9 @@ typedef struct {
 	uint8_t          load_balance;
 	uint16_t         max_spread;
 	uint32_t         ring_mask;
-	odp_spinlock_t   mask_lock;
 	odp_atomic_u32_t grp_epoch;
 	odp_shm_t        shm;
+	odp_spinlock_t   mask_lock[NUM_SCHED_GRPS];
```
Comment (MatiasElo) [inline platform/linux-generic/odp_schedule_basic.c]: Typo in the commit message: `... which is called from in the fast path ...`

File: platform/linux-generic/odp_schedule_basic.c [hunk 4]
```diff
@@ -22,7 +22,7 @@
 #include <odp/api/plat/thread_inlines.h>
 #include <odp/api/time.h>
 #include <odp/api/plat/time_inlines.h>
-#include <odp/api/spinlock.h>
+#include <odp/api/ticketlock.h>
```
Comment (MatiasElo) [inline platform/linux-generic/odp_schedule_basic.c:25]: Commit message: `Ticket lock should may give ...`

**Review by MatiasElo — APPROVED** [review]

File: platform/linux-generic/odp_schedule_basic.c (hunk 4 above)
Comment (psavol) [reply platform/linux-generic/odp_schedule_basic.c:25]: fixed in v6

File: platform/linux-generic/odp_schedule_basic.c (hunk 3 above)
Comment (psavol) [reply platform/linux-generic/odp_schedule_basic.c]: fixed in v6

**Pull request merged**